### PR TITLE
Make RestHighLevelClient Closeable and simplify its creation

### LINKED ISSUE #26086: Merge High/Low Level REST client functionalities
When using the high level REST client, one would need to create a low level REST client to manage connection options.  The high level client is only used to serialize and deserialize JSON requests and responses.  It seems the high level REST client is an aux pack to the low level REST client.  It is cumbersome to use two "clients" (high/low).  It would be simpler just to have a single client (encompassing high and low) to create connections, query, index, and make some cluster wide calls like cluster health.

### PR DESCRIPTION
**Pull request by javanna**
By making RestHighLevelClient Closeable, its close method will close the internal low-level REST client instance, which simplifies the way users interact with the high-level client as they no longer need to keep track of the low-level client provided at initialization time.

Also, a new constructor that accepts HttpHosts is added to RestHighLevelClient, so that there is a way to create such client without specifying the underlying RestClient instance. The advantage is that users do not always have to create two clients. The disadvantage is that the dependency on apache http client is exposed through this constructor, but that is already exposed anyway by the low-level client which the high-level client uses.

Closes #26086

### CODE REVIEW

**Comment by dadoonet** [comment]
Can we have also nullable CredentialsProvider parameter in the CTOR?

**Comment by javanna** [comment]
> Can we have also nullable CredentialsProvider parameter in the CTOR?

I wouldn't go that route. The new constructor is to facilitate the simplest usecase. I expect users to understand that we create a client with default options for them in that case, and that if they need to play with http settings they have to provide an externally created low-level REST client. If the outcome of adding this new constructor is adding more arguments to it I am leaning towards not adding it at all. @cbuescher @tlrx what do you think?

**Comment by dadoonet** [comment]
I would expect our users using xpack. Having only the CTOR without a way to pass credentials makes less sense to me.

**Review by cbuescher — APPROVED** [review]
LGTM

**Review by tlrx — APPROVED** [review]
LGTM

**Comment by nik9000** [comment]
I kind of don't like the new ctor. I love closing the low level client when closing the high level client but I like how "obvious" having only a single public ctor for the high level client is. With two it just feels less obvious. I dunno.

**Comment by jasontedor** [comment]
I think it's abnormal for the high-level client to take ownership of the lifecycle of the provided low-level client. Maybe that's what @nik9000 means by it being less obvious? Either way, I'm opposed to the approach here.

**Comment by nik9000** [comment]
> I think it's abnormal for the high-level client to take ownership of the lifecycle of the provided low-level client.

I think of making the high level client `Closeable` and pushing the close to the low level client as similar to how `InputStreamReader#close` closes the underlying stream. So I don't mind the lifecycle. I just liked that we had one constructor and it was "obvious" that the high level client delegated to the low level client because that is how you built the thing. I feel like this is a good abstraction to leak because the low level client has a lot of interesting things to configure.

**Comment by javanna** [comment]
I tend to agree with @nik9000 on the new constructor. I added it because I heard some users confused about having to create two clients, and thought that in the simplest case we can create it for them, but I would much rather prefer users to understand the difference between low-level and high-level. After all it may just be a naming issue. I think I will remove the additional constructor.

@jasontedor do you have a better approach in mind? The problem we are trying to solve is that some users are creating their own wrapper because they want to interact with a single object, while we require to send requests through high-level client but close the low-level one. I do see what you mean though: "if you provide the low-level client yourself, why would we close it for you?". Seems like for some users keeping track of both clients after creation complicates things, see previous discussion in #26086 .

**Comment by rjernst** [comment]
To me this is like Readers/InputStreams in java. If you call `close()` on the outer wrapper, close is called all the way to the innermost stream.  By passing the low level client to the high level client, you are giving up ownership of it.

Perhaps a different way to make this clear would be to allow the high level client to construct the low level client indirectly, by passing a Supplier to the HL ctor.  Then, having two constructors, one that is the "simple" one which constructs the LL client, and one that takes the Supplier so you can build the client yourself, but then clearly do not own it since you did not construct it.

**Comment by nik9000** [comment]
One wrinkle is that, at least for now, I frequently want to use the low level client for things that the high level client doesn't support. If the high level client had a method that let me fetch the low level client from it then I'd be happy with the indirect ctor approach.

**Comment by nik9000** [comment]
But, yeah, "unwrapping" the high level client does indeed make this less like Readers and Streams.

**Comment by javanna** [comment]
I like the `Supplier` idea. We could add a getter to retrieve the underlying low-level client instance.

I still wonder if we really want a "simpler" constructor. If we add one that only accepts the hosts, users may ask for more arguments like credentials etc. also I think that most users are going to use the "advanced" constructor in practice, so maybe it is not even worth adding the simple one. Thoughts?

Would we still make the supplier change if we keep one public constructor only? I'd think so.

**Comment by nik9000** [comment]
I agree with having a single public ctor. I think passing in the low level client is ok, but if we really want the high level client to own the low level one we could make it take `RestClientBuilder` in its ctor. So it'd be:
```
RestHighLevelClient client = new RestHighLevelClient(RestClient.builder(host, host host)
    .setMaxRetryTimeoutMillis(1000)
    .setHttpClientConfigCallback(config -> {
        config.setProxy(proxy);
    }));
```

I like this because the method signatures tell you what you can do. You have to know `RestHighLevelClient` but after that it tells you you need a `RestClientBuilder`. We ought to have in the class javadoc for that class that the way to get it is `RestClient.builder`.

It'd be a little more obvious if it were:
```
RestHighLevelClient client = new RestHighLevelClient(new HttpHost[] {host, host host}, builder -> 
    builder.setMaxRetryTimeoutMillis(1000)
        .setHttpClientConfigCallback(config -> {
           config.setProxy(proxy);
        }));
```

but I don't think it'd be *better*.

**Review by cbuescher — COMMENTED** [review]
After all this discussion, I think adding the second ctor in this PR is not the best choice, also using `RestClientBuilder` to me seems more confusing than simply supplying a low level client build externally. Currently people will also need a reference to the LL client occasionally to issue admin request we don't support in the HL client yet. I think keeping an external reference to the LL client currently is better than addding a getter for it to the HL client.

**Comment by jasontedor** [comment]
> I think keeping an external reference to the LL client currently is better than addding a getter for it to the HL client.

And this is why the high-level client should not take ownership. The low-level client has a lifecycle useful outside of the lifecycle of the high-level client. It's not a like an input stream where an input stream reader really is taking ownership of the underlying stream. Once the reader is done consuming the stream, the stream is done. And it typically can not be shared like the low-level client can.

**Comment by dadoonet** [comment]
As I said previously IMHO the HL Client extends features provided by the Low Level one.

That's why I think it should simply extend in Java the LowLevel class.

Then you don't need to close both but just the HL one (which is actually the Low level one).

**Comment by javanna** [comment]
As much as I'd love to make everybody, opinions in this discussion are quite the opposite. I understand both sides and I don't have strong opinions in favour or against either of them. What matters the most to me is that things are as clear as possible to users and that we don't force users to create their own wrappers/abstractions.

I take it that the change that most people (besides Jason if I understood correctly) find reasonable is making `RestHighLevelClient` `Closeable`, without adding any constructor nor changing the existing one. Users would still have to provide an externally built low-level `RestClient` but after construction they no longer have to keep track of it. 

The supplier idea (or `RestClientBuilder`) requires adding a getter that some people are against to. I would rather try and clarify through javadoc that closing the high-level client is enough. One can still close the low-level one externally (and even if by mistake it gets closed twice, nothing happens).

This simple change would help users that currently end up creating their own wrapper class so that they interact with a single client abstraction. It is true that they would just need to keep track of two references instead but I think that the expectation of interacting with a single client object is reasonable and we could do something to address this.

**Comment by javanna** [comment]
I pushed a new commit. We had some more internal discussions and I've come up with the following solution which takes into account all the feedback/opinions:

- Make `RestHighLevelClient` accept `RestClientBuilder` as a constructor argument. Most users will use this constructor. Passing the builder rather than an already built client clarifies who creates the object, and who closes it as well. This is a compromise between adding a constructor that accepts hosts only and keeping a public constructor that accepts an externally created low-level client for cases where connections settings etc. need to be customized (very common). This is also the only exposed public constructor which allows to keep things simple.

- Make `RestHighLevelClient` `Closeable`. The `close` method will always close the internal low-level client implementation, which is though internally created so that is less of a surprise with the above change.

- Leave a protected constructor that accepts a `RestClient` instance. This can be used by subclasses. Subclassing is required for advanced usecases where for instance an externally built low-level client needs to be provided. In such case it is also possible to override the close method and make it a no-op so that the lifecycle of the provided low-level client is externally managed.

- add a `getLowLevelClient` method to retrieve the internally created low-level client. This is  common given that the high-level client supports only some of the Elasticsearch APIs, and we don't want users to have to create an additional low-level client instance if they want to use endpoints that are not yet supported. They can then just retrieve the existing one and use it.

- given that we allow to retrieve the internal low-level client, add some protection to prevent the low-level client from being closed multiple times.

**Comment by dadoonet** [comment]
Awesome 👏. Thanks!

**Comment by rjernst** [comment]
> add some protection to prevent the low-level client from being closed multiple times.

`close()` needs to be idemptotent. From the javadocs:

```If the stream is already closed then invoking this method has no effect.```

**Comment by javanna** [comment]
> close() needs to be idemptotent.

cool I will remove the protection then. It is idempotent as-is.

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java [hunk 1]
```diff
@@ -185,6 +193,18 @@ protected RestHighLevelClient(RestClient restClient, List<NamedXContentRegistry.
...
+     * Returns the low-level client that the current high-level client instance is using to perform requests
+     */
+    public RestClient getLowLevelClient() {
+        return client;
+    }
+
+    @Override
+    public void close() throws IOException {
+        client.close();
+    }
```
Comment (cbuescher) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java]: If its not complicated, maybe add a test that HL close also closes the LL client and is idempotent (as of last change requested by Ryan).

**Review by cbuescher — APPROVED** [review]
Still LGTM, I left one small suggestion for an additional test, since everything revolves around closing or not closing...

**Comment by jasontedor** [comment]
This is sounding more appealing to me @javanna. Although, I wonder what you think of the following suggestion:

```diff
diff --git a/client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java b/client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java
index 5d66618d5d..5a1c5ff875 100644
--- a/client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java
+++ b/client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java
@@ -43,6 +43,7 @@ import org.elasticsearch.action.update.UpdateRequest;
 import org.elasticsearch.action.update.UpdateResponse;
 import org.elasticsearch.client.http.Header;
 import org.elasticsearch.client.http.HttpEntity;
+import org.elasticsearch.common.CheckedConsumer;
 import org.elasticsearch.common.CheckedFunction;
 import org.elasticsearch.common.ParseField;
 import org.elasticsearch.common.xcontent.ContextParser;
@@ -161,7 +162,7 @@ import static java.util.stream.Collectors.toList;
  * The {@link RestClient} instance is internally built based on the provided {@link RestClientBuilder} and it gets closed automatically
  * when closing the {@link RestHighLevelClient} instance that wraps it.
  * In case an already existing instance of a low-level REST client needs to be provided, this class can be subclassed and the
- * {@link #RestHighLevelClient(RestClient, List)} constructor can be used. Also, the {@link #close()} method can be overridden with an
+ * {@link #RestHighLevelClient(RestClient, List, CheckedConsumer)} constructor can be used. Also, the {@link #close()} method can be overridden with an
  * empty implementation if it is preferable to externally close the provided low-level REST client.
  * This class can also be sub-classed to expose additional client methods that make use of endpoints added to Elasticsearch through
  * plugins, or to add support for custom response sections, again added to Elasticsearch through plugins.
@@ -170,26 +171,25 @@ public class RestHighLevelClient implements Closeable {
 
     private final RestClient client;
     private final NamedXContentRegistry registry;
+    private final CheckedConsumer<RestClient, IOException> doClose;
 
     /**
      * Creates a {@link RestHighLevelClient} given the low level {@link RestClientBuilder} that allows to build the
      * {@link RestClient} to be used to perform requests.
      */
     public RestHighLevelClient(RestClientBuilder restClientBuilder) {
-        this(restClientBuilder.build(), Collections.emptyList());
+        this(restClientBuilder.build(), Collections.emptyList(), RestClient::close);
     }
 
-    /**
-     * Creates a {@link RestHighLevelClient} given the low level {@link RestClient} that it should use to perform requests and
-     * a list of entries that allow to parse custom response sections added to Elasticsearch through plugins.
-     * This constructor can be called by subclasses in case an externally created low-level REST client needs to be provided.
-     * Also subclasses can provide parsers for custom response sections added to Elasticsearch through plugins.
-     */
-    protected RestHighLevelClient(RestClient restClient, List<NamedXContentRegistry.Entry> namedXContentEntries) {
+    protected RestHighLevelClient(
+            final RestClient restClient,
+            final List<NamedXContentRegistry.Entry> namedXContentEntries,
+            final CheckedConsumer<RestClient, IOException> doClose) {
         this.client = Objects.requireNonNull(restClient);
         this.registry = new NamedXContentRegistry(
                 Stream.of(getDefaultNamedXContents().stream(), getProvidedNamedXContents().stream(), namedXContentEntries.stream())
-                    .flatMap(Function.identity()).collect(toList()));
+                        .flatMap(Function.identity()).collect(toList()));
+        this.doClose = doClose;
     }
 
     /**
@@ -200,8 +200,8 @@ public class RestHighLevelClient implements Closeable {
     }
 
     @Override
-    public void close() throws IOException {
-        client.close();
+    public final void close() throws IOException {
+        doClose.accept(client);
     }
 
     /**
```

Now this makes it clear that if you extend you have to make an explicit decision on whether or not the HL client owns the LL client.

**Comment by javanna** [comment]
@jasontedor I integrated your proposal, I like how explicit it makes the decision to be made.

**Review by jasontedor — APPROVED** [review]
LGTM!

File: client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java [hunk 2]
```diff
@@ -157,34 +159,64 @@
...
      * a list of entries that allow to parse custom response sections added to Elasticsearch through plugins.
+     * This constructor can be called by subclasses in case an externally created low-level REST client needs to be provided.
+     * The consumer argument allows to control what needs to be done when the {@link #close()} method is called.
+     * Also subclasses can provide parsers for custom response sections added to Elasticsearch through plugins.
      */
-    protected RestHighLevelClient(RestClient restClient, List<NamedXContentRegistry.Entry> namedXContentEntries) {
+    protected RestHighLevelClient(RestClient restClient, CheckedConsumer<RestClient, IOException> doClose,
+                                  List<NamedXContentRegistry.Entry> namedXContentEntries) {
         this.client = Objects.requireNonNull(restClient);
+        this.doClose = doClose;
```
Comment (tlrx) [inline client/rest-high-level/src/main/java/org/elasticsearch/client/RestHighLevelClient.java]: ` this.doClose = Objects.requireNonNull(doClose);` too?

**Pull request merged**